javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,303 @@
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.detach()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getUnregistered() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getClosed() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).close()
+    verify(writeMock, times(1)).close()
+  }
+
+  @Test
+  fun reset_allHandles_allElementsRemoved() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    handleHolder.setHandle("readHandle", mock())
+    handleHolder.setHandle("writeHandle", mock())
+    assertThat(handleHolder.handles).isNotEmpty()
+
+    handleHolder.reset()
+
+    assertThat(handleHolder.handles).isEmpty()
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap_returnsTrue() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    assertThat(handleHolder.isEmpty()).isTrue()
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap_returnsFalse() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    handleHolder.setHandle("readHandle", mock())
+
+    assertThat(handleHolder.isEmpty()).isFalse()
+  }
+
+  @Test
+  fun createForeignReference_noHandles_asserts() = runBlockingTest {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    assertFailsWith<IllegalStateException> {
+      handleHolder.createForeignReference<EntityBase>(mock(), "someId")
+    }
+  }
+
+  @Test
+  fun createForeignReference_someHandles_success() = runBlockingTest {
+    class DummyHandle : HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    ) {
+      val readHandle: Handle by handles
+    }
+
+    val handleHolder = DummyHandle()
+    val mock = mock<Handle>()
```

Comment:
maybe name this variable readHandleMock (to clearly distinguish between it and the entitySpec handle mock)?